src/mlpack/methods/preprocess/preprocess_string_main.cpp

```diff
@@ -0,0 +1,301 @@
+                   const std::string& filename,
+                   const std::string& column_delimiter)
+{
+  // Extracting the Contents of file
+  // File pointer
+  ifstream fin;
+  // Open an existing file
+  fin.open(filename);
+  if (!fin.is_open())
+  {
+    Log::Fatal << "Unable to open input file \n";
+  }
+  std::string line, word;
+  boost::string_view token, copy;
+  while (std::getline(fin, line))
+  {
+    stringstream streamLine(line);
+    dataset.emplace_back(std::vector<std::string>());
+    // delimeter[0] becase the standard function accepts char as input.
+    while (std::getline(streamLine, word, column_delimiter[0]))
+    {
+      dataset.back().push_back(word);
+    }
+  }
+}
+
+void checkForDigit(const std::string& column)
+{
+  for (auto& i : column)
+    if (!std::isdigit(i))
+    Log::Fatal << "Characters encountered, Please give only digits";
+}
+
+void getColumn(const std::vector<std::string>& temp_dimension,
+               std::set<size_t>& dimension)
+{
+  int columnstartindex, columnendindex;
+  size_t found;
+  for (size_t i = 0; i < temp_dimension.size(); i++)
+  {
+    found = temp_dimension[i].find('-');
+    if ( found != string::npos)
+    {
+      try
+      {
+        // Has a range include, something of type a-b.
+        checkForDigit(temp_dimension[i].substr(0, found));
+        checkForDigit(temp_dimension[i].substr(found+1,
+            temp_dimension[i].length()));        
+        columnstartindex = std::stoi(temp_dimension[i].substr(0, found));
+        columnendindex = std::stoi(temp_dimension[i].substr(found+1,
+            temp_dimension[i].length()));
+      }
+      catch (const std::exception& e)
+      {
+        Log::Fatal << "Dimension value not clear, either negatve or can't "
+        "parse the range. Usage a-b \n";
+      }
+      for (int i = columnstartindex; i <= columnendindex; i++)
+        dimension.insert(i);
+    }
+    else
+    {
+      try
+      {
+        checkForDigit(temp_dimension[i]);
+        dimension.insert(std::stoi(temp_dimension[i]));
+      }
+      catch (const std::exception& e)
+      {
+        Log::Fatal << "Dimension value not appropriate \n";      
+      }
+    }
+  }
+}
+
+void writeOutput(const std::string& outputFilename,
+                 const std::vector<std::vector<std::string>>& dataset,
+                 const std::vector<std::vector<std::string>>& nonNumericInput,
+                 const std::string& column_delimiter,
+                 const std::set<size_t>& dimension)
+{
+  ofstream fout;
+  fout.open(outputFilename, ios::trunc);
+  if (!fout.is_open())
+  {
+    Log::Fatal << "Unable to open a file for writing output.\n";
+  }
+  std::set<size_t>::iterator dimensionIterator;
+  for (size_t i = 0 ; i < dataset.size(); i++)
+  {
+    size_t col = 0;
+    dimensionIterator = dimension.begin();
+    for (size_t j = 0 ; j < dataset[i].size(); j++)
+    {
+      if (col < dimension.size() && *dimensionIterator == j)
```

Comment:
Again, It's too complecated. You can use `unordered_set` instead of `set`:
```
if (dimenstions.find(j) != dimensions.end())
```

The complexity won't change.